backpack/core/derivatives/batchnorm_nd.py

```diff
@@ -0,0 +1,327 @@
+            equation = "vncl,c->vncl".replace("l", self.free_index[N])
+            jac_t_mat = einsum(equation, jac_t_mat, ivar / denominator)
+            return jac_t_mat
+        else:
+            equation = "c,vncl->vncl".replace("l", self.free_index[N])
+            return einsum(
+                equation,
+                ((module.running_var + module.eps) ** (-0.5)) * module.weight,
+                mat,
+            )
+
+    def _weight_jac_mat_prod(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+    ) -> Tensor:
+        x_hat, _ = self._get_normalized_input_and_var(module)
+        equation: str = "ncl,vc->vncl".replace(
+            "l", self.free_index[self._get_n_axis(module)]
+        )
+        return einsum(equation, x_hat, mat)
+
+    def _weight_jac_t_mat_prod(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+        sum_batch: bool = True,
+    ) -> Tensor:
+        x_hat, _ = self._get_normalized_input_and_var(module)
+        equation = f"vncl,ncl->v{'' if sum_batch else 'n'}c".replace(
+            "l", self.free_index[self._get_n_axis(module)]
+        )
+        return einsum(equation, mat, x_hat)
+
+    def _bias_jac_mat_prod(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+    ) -> Tensor:
+        out = self._unsqueeze_free_axis(module, mat, 1)
+        dim_expand: List[int] = [-1, module.input0.shape[0], -1]
+        for n in range(self._get_n_axis(module)):
+            dim_expand.append(module.input0.shape[2 + n])
+        return out.expand(*dim_expand)
+
+    def _bias_jac_t_mat_prod(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+        sum_batch: bool = True,
+    ) -> Tensor:
+        axis_sum: Tuple[int] = self._get_free_axes(module, with_batch_axis=sum_batch)
+        return mat.sum(dim=axis_sum) if axis_sum else mat
+
+    def _residual_mat_prod(
+        self,
+        module: BatchNorm1d,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+    ) -> Tensor:
+        """Multiply with BatchNorm1d residual-matrix.
+
+        Paul Fischer (GitHub: @paulkogni) contributed this code during a research
+        project in winter 2019.
+
+        Details are described in
+
+        `HESSIAN BACKPROPAGATION FOR BATCHNORM`
+        <https://uni-tuebingen.de/en/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/methoden-des-maschinellen-lernens/personen/alumni/everyone-else/>
+        <TODO: Wait for tech report upload and add exact link>_
```

Comment:
Waiting for the upload. Should be done until Monday.

Reply:
Document is up: https://uni-tuebingen.de/securedl/sdl-eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpYXQiOjE2MjU0ODY0NzAsImV4cCI6MTYyNTU3NjQ2NCwidXNlciI6MCwiZ3JvdXBzIjpbMCwtMV0sImZpbGUiOiJmaWxlYWRtaW5cL1VuaV9UdWViaW5nZW5cL0Zha3VsdGFldGVuXC9NYXROYXRcL0ZhY2hiZXJlaWNoZVwvSW5mb3JtYXRpa1wvTGVocnN0dWVobGVcL01ldGhNYXNjaExlcm5cL0Rva3VtZW50ZVwvVGhlc2VzXC9BYnNjaGx1c3NiZXJpY2h0X0Zpc2NoZXIucGRmIiwicGFnZSI6MTczNDc1fQ.xMe4WmTsyak9J-C7iTxGqTdpYMxMxtGfouyAZgW158I/Abschlussbericht_Fischer.pdf

Reply:
Link inserted.